File: app/src/main/java/com/dublikunt/nclient/api/components/GenericGallery.java
package com.dublikunt.nclient.api.components;

import android.os.Parcelable;

import androidx.annotation.NonNull;

import com.dublikunt.nclient.components.classes.Size;
import com.dublikunt.nclient.files.GalleryFolder;
import com.dublikunt.nclient.utility.Utility;

import java.util.Locale;

public abstract class GenericGallery implements Parcelable {

    public abstract int getId();

    public abstract Type getType();

    public abstract int getPageCount();

    public abstract boolean isValid();

    @NonNull
    public abstract String getTitle();

    public abstract Size getMaxSize();

    public abstract Size getMinSize();

    public abstract GalleryFolder getGalleryFolder();

    public String sharePageUrl(int i) {
        return String.format(Locale.US, "https://" + Utility.getHost() + "/g/%d/%d/", getId(), i + 1);
    }

    public boolean isLocal() {
        return getType() == Type.LOCAL;
    }

    public abstract boolean hasGalleryData();

    public abstract GalleryData getGalleryData();

    public enum Type {COMPLETE, LOCAL, SIMPLE}

}

File: app/src/main/kotlin/com/dublikunt/nclient/api/components/GenericGallery.kt
package com.dublikunt.nclient.api.components

import android.os.Parcelable
import com.dublikunt.nclient.classes.Size
import com.dublikunt.nclient.files.GalleryFolder
import com.dublikunt.nclient.utility.Utility
import java.util.*

abstract class GenericGallery : Parcelable {
    abstract val id: Int
    abstract val type: Type
    abstract val pageCount: Int
    abstract val valid: Boolean
    abstract val title: String
    abstract val maxSize: Size?
    abstract val minSize: Size?
    abstract val galleryFolder: GalleryFolder?
    abstract val galleryData: GalleryData?
    val isLocal: Boolean
        get() = type == Type.LOCAL

    abstract fun hasGalleryData(): Boolean

    enum class Type {
        COMPLETE, LOCAL, SIMPLE
    }

    fun sharePageUrl(i: Int): String {
        return String.format(Locale.US, "https://" + Utility.host + "/g/%d/%d/", id, i + 1)
    }
}
